## Phase 2 (Data Understanding): Ausreißeranalyse

* Autorin: Anna (i3-Versicherung)
* Webseite: [Data Science Training - Kapitel 2](https://data-science.training/kapitel-2/)
* Datum: 23.03.2023

Wir führen die Trainings- und Testdaten zusammen und suchen nach (numerischen) Ausreißern der Attribute Fare, Age und SibSp. Dabei hilft uns die Standardnormalverteilung (mit Mittelwert 0 und Standardabweichung 1). Wenn ein Z-skalierter Attributwert größer als die n-fache Standardabweichung ist, gilt er als Ausreißer. Bei negativen Werten benutzen wir "kleiner als" statt "größer als". Mit Hilfe des Absolutbetrags (abs) können wir beide Fälle auch gemeinsam betrachten. Der Parameter n = 4 liefert meistens gute Ergebnisse.

In [4]:
# Pandas Paket (Package) importieren
#  Datenstrukturen und Datenanalyse, I/O
#  https://pandas.pydata.org/pandas-docs/stable/
import pandas as pd
# NumPy Paket (Package) importieren
#  Mehrdimensionale Datenstrukturen (Vektoren, Matrizen, Tensoren, Arrays), Lineare Algebra
#  https://numpy.org/doc/
import numpy as np

In [5]:
# Trainings- und Testdaten als Pandas Data Frame (df) aus CSV-Dateien laden
#  (KNIME: "CSV Reader")
df_train = pd.read_csv('../../data/titanic/original/train.csv')
df_test  = pd.read_csv('../../data/titanic/original/test.csv')

In [6]:
# Daten zusammenführen
#  (KNIME "Concatenate")
df = pd.concat([df_train, df_test], ignore_index=True)

In [7]:
# Datentypen automatisch konvertieren
#  Diese Umwandlung macht später Probleme, wenn die fehlenden Werte zuvor nicht behandelt wurden.
#df = df.convert_dtypes()

In [8]:
# Funktion (Definition): Den kleinsten (positiven) Ausreißer bestimmen
#  (KNIME: "Numeric Outliers")
# 
# In Anlehnung an:
# https://www.kaggle.com/code/rhythmcam/titanic-zscore-outlier-detection
#
def dst_zscore_outlier(df, col, n=4):
    
    # Initialisierung und Berechnung von Mittelwert und Standardabweichung
    out = [] # Leere Liste
    mn  = np.mean(df[col]) # Mittelwert
    sd  = np.std (df[col]) # Standardabweichung
    
    # Schleife über alle Datenzeilen
    for x in df[col]: 
        z = (x - mn) / sd # z-score (so dass gilt: Mittelwert = 0, Standardabweichung = 1)
        if np.abs(z) > n: # z-score gößer als die n-fache Standardabweichung ?
            out.append(x) # Ausreißer zur Liste hinzufügen
    
    # Rückgabe: Minimum der Ausreißerwerte
    if out:
        return np.min(out)
    else:
        return np.inf # Unendlich bei leerer Liste

In [9]:
# Funktion (Definition): Alle (positiven) Ausreißer bestimmen und ausgeben
#
def dst_outliers(df, col, n=4):
    
    # 1. Den kleinsten Ausreißer bestimmen
    min_outlier = dst_zscore_outlier(df, col, n)
    
    # 2. Nach allen Ausreißern filtern
    df1 = df[df[col] >= min_outlier]
    
    # 3. Die Daten absteigend sortieren (d.h. die größten Ausreißer kommen zuerst)
    df1 = df1.sort_values(by=col, ascending=False)
    
    # 4. Die Anzahl der Ausreißer ausgeben
    print(col, ': Anzahl Ausreißer =', df1.shape[0])
    
    # 5. Die Tabelle mit allen Ausreißern anzeigen
    display(df1)

In [10]:
# Fare
dst_outliers(df, 'Fare')

Fare : Anzahl Ausreißer = 20


PassengerId  Survived  Pclass  \
1234         1235       NaN       1   
258           259       1.0       1   
679           680       1.0       1   
737           738       1.0       1   
27             28       0.0       1   
960           961       NaN       1   
341           342       1.0       1   
438           439       0.0       1   
88             89       1.0       1   
944           945       NaN       1   
950           951       NaN       1   
1033         1034       NaN       1   
955           956       NaN       1   
742           743       1.0       1   
915           916       NaN       1   
311           312       1.0       1   
1266         1267       NaN       1   
299           300       1.0       1   
1075         1076       NaN       1   
118           119       0.0       1   

                                                   Name     Sex   Age  SibSp  \
1234  Cardeza, Mrs. James Warburton Martinez (Charlo...  female  58.0      0   
258                                    Ward, Miss. Anna  female  35.0      0   
679                  Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
737                              Lesurer, Mr. Gustave J    male  35.0      0   
27                       Fortune, Mr. Charles Alexander    male  19.0      3   
960                 Fortune, Mrs. Mark (Mary McDougald)  female  60.0      1   
341                      Fortune, Miss. Alice Elizabeth  female  24.0      3   
438                                   Fortune, Mr. Mark    male  64.0      1   
88                           Fortune, Miss. Mabel Helen  female  23.0      3   
944                          Fortune, Miss. Ethel Flora  female  28.0      3   
950                         Chaudanson, Miss. Victorine  female  36.0      0   
1033                         Ryerson, Mr. Arthur Larned    male  61.0      1   
955                         Ryerson, Master. John Borie    male  13.0      2   
742               Ryerson, Miss. Susan Parker "Suzette"  female  21.0      2   
915     Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0      1   
311                          Ryerson, Miss. Emily Borie  female  18.0      2   
1266                           Bowen, Miss. Grace Scott  female  45.0      0   
299     Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
1075  Douglas, Mrs. Frederick Charles (Mary Helene B...  female  27.0      1   
118                            Baxter, Mr. Quigg Edmond    male  24.0      0   

      Parch    Ticket      Fare            Cabin Embarked  
1234      1  PC 17755  512.3292      B51 B53 B55        C  
258       0  PC 17755  512.3292              NaN        C  
679       1  PC 17755  512.3292      B51 B53 B55        C  
737       0  PC 17755  512.3292             B101        C  
27        2     19950  263.0000      C23 C25 C27        S  
960       4     19950  263.0000      C23 C25 C27        S  
341       2     19950  263.0000      C23 C25 C27        S  
438       4     19950  263.0000      C23 C25 C27        S  
88        2     19950  263.0000      C23 C25 C27        S  
944       2     19950  263.0000      C23 C25 C27        S  
950       0  PC 17608  262.3750              B61        C  
1033      3  PC 17608  262.3750  B57 B59 B63 B66        C  
955       2  PC 17608  262.3750  B57 B59 B63 B66        C  
742       2  PC 17608  262.3750  B57 B59 B63 B66        C  
915       3  PC 17608  262.3750  B57 B59 B63 B66        C  
311       2  PC 17608  262.3750  B57 B59 B63 B66        C  
1266      0  PC 17608  262.3750              NaN        C  
299       1  PC 17558  247.5208          B58 B60        C  
1075      1  PC 17558  247.5208          B58 B60        C  
118       1  PC 17558  247.5208          B58 B60        C

In [11]:
# Age
dst_outliers(df, 'Age', 2.5) # hier 2,5 (statt 4) Standardabweichungen als maximal erlaubte Abweichung

Age : Anzahl Ausreißer = 10


PassengerId  Survived  Pclass  \
630          631       1.0       1   
987          988       NaN       1   
851          852       0.0       3   
96            97       0.0       1   
493          494       0.0       1   
116          117       0.0       3   
672          673       0.0       2   
745          746       0.0       1   
972          973       NaN       1   
33            34       0.0       2   

                                                  Name     Sex   Age  SibSp  \
630               Barkworth, Mr. Algernon Henry Wilson    male  80.0      0   
987  Cavendish, Mrs. Tyrell William (Julia Florence...  female  76.0      1   
851                                Svensson, Mr. Johan    male  74.0      0   
96                           Goldschmidt, Mr. George B    male  71.0      0   
493                            Artagaveytia, Mr. Ramon    male  71.0      0   
116                               Connors, Mr. Patrick    male  70.5      0   
672                        Mitchell, Mr. Henry Michael    male  70.0      0   
745                       Crosby, Capt. Edward Gifford    male  70.0      1   
972                                 Straus, Mr. Isidor    male  67.0      1   
33                               Wheadon, Mr. Edward H    male  66.0      0   

     Parch      Ticket      Fare    Cabin Embarked  
630      0       27042   30.0000      A23        S  
987      0       19877   78.8500      C46        S  
851      0      347060    7.7750      NaN        S  
96       0    PC 17754   34.6542       A5        C  
493      0    PC 17609   49.5042      NaN        C  
116      0      370369    7.7500      NaN        Q  
672      0  C.A. 24580   10.5000      NaN        S  
745      1   WE/P 5735   71.0000      B22        S  
972      0    PC 17483  221.7792  C55 C57        S  
33       0  C.A. 24579   10.5000      NaN        S

In [12]:
# SibSp
dst_outliers(df, 'SibSp')

SibSp : Anzahl Ausreißer = 15


PassengerId  Survived  Pclass                                Name  \
159           160       0.0       3          Sage, Master. Thomas Henry   
180           181       0.0       3        Sage, Miss. Constance Gladys   
201           202       0.0       3                 Sage, Mr. Frederick   
324           325       0.0       3            Sage, Mr. George John Jr   
792           793       0.0       3             Sage, Miss. Stella Anna   
846           847       0.0       3            Sage, Mr. Douglas Bullen   
863           864       0.0       3   Sage, Miss. Dorothy Edith "Dolly"   
1079         1080       NaN       3                     Sage, Miss. Ada   
1251         1252       NaN       3         Sage, Master. William Henry   
59             60       0.0       3  Goodwin, Master. William Frederick   
71             72       0.0       3          Goodwin, Miss. Lillian Amy   
386           387       0.0       3     Goodwin, Master. Sidney Leonard   
480           481       0.0       3      Goodwin, Master. Harold Victor   
683           684       0.0       3         Goodwin, Mr. Charles Edward   
1031         1032       NaN       3         Goodwin, Miss. Jessie Allis   

         Sex   Age  SibSp  Parch    Ticket   Fare Cabin Embarked  
159     male   NaN      8      2  CA. 2343  69.55   NaN        S  
180   female   NaN      8      2  CA. 2343  69.55   NaN        S  
201     male   NaN      8      2  CA. 2343  69.55   NaN        S  
324     male   NaN      8      2  CA. 2343  69.55   NaN        S  
792   female   NaN      8      2  CA. 2343  69.55   NaN        S  
846     male   NaN      8      2  CA. 2343  69.55   NaN        S  
863   female   NaN      8      2  CA. 2343  69.55   NaN        S  
1079  female   NaN      8      2  CA. 2343  69.55   NaN        S  
1251    male  14.5      8      2  CA. 2343  69.55   NaN        S  
59      male  11.0      5      2   CA 2144  46.90   NaN        S  
71    female  16.0      5      2   CA 2144  46.90   NaN        S  
386     male   1.0      5      2   CA 2144  46.90   NaN        S  
480     male   9.0      5      2   CA 2144  46.90   NaN        S  
683     male  14.0      5      2   CA 2144  46.90   NaN        S  
1031  female  10.0      5      2   CA 2144  46.90   NaN        S

### Ergebnis der Ausreißeranalyse

Fare
* 20 Passagiere haben einen Ticket-Preis von mehr als 240 Pfund bezahlt.
* Insbesondere Fam. Cardeza sticht heraus (Luxus-Kabinen B51 B53 B55).
* 17 Passagiere haben nicht für die Überfahrt bezahlen müssen.

Age
* 9 Passagiere sind 67 Jahre oder älter.

SibSp
* Die Großfamilien Sage und Goodwin fallen hier auf.

In [14]:
# B51 B53 B55: Was sind das für extrem teure Kabinen?
# Wer hat diese belegt?
df_check = df[df['Cabin'] == 'B51 B53 B55'] 
display(df_check)

PassengerId  Survived  Pclass  \
679           680       1.0       1   
872           873       0.0       1   
1234         1235       NaN       1   

                                                   Name     Sex   Age  SibSp  \
679                  Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
872                            Carlsson, Mr. Frans Olof    male  33.0      0   
1234  Cardeza, Mrs. James Warburton Martinez (Charlo...  female  58.0      0   

      Parch    Ticket      Fare        Cabin Embarked  
679       1  PC 17755  512.3292  B51 B53 B55        C  
872       0       695    5.0000  B51 B53 B55        S  
1234      1  PC 17755  512.3292  B51 B53 B55        C

### Inkonsistente (fehlerhafte Daten)

Der Passagier Mr. Frans Olof Carlsson (PassengerId = 873) hat die gleiche Kabinennummer (B51 B53 B55) wie die Fam. Cardeza (s.o.). Er ist allein gereist und hat nur 5 Pfund bezahlt.